### PR TITLE
Balloons: add e2e tests for topology aware allocation and idle CPU sharing

### PR DESCRIPTION
**Pull request by fmuyassarov**
This PR extends E2E to test idle CPU sharing (`ShareIdleCpusInSame` field)
and topology aware creation/inflation/deflation that is implemented in https://github.com/intel/cri-resource-manager/pull/931.

### CODE REVIEW

File: test/e2e/policies.test-suite/balloons/n4c32/test01-dynamic-baloons/code.var.sh [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+# pod0-pod7: create 8 balloons, where each lands on a different NUMA node.
+# Each balloon (except one that lands on the NUMA node with reserved CPUs)
+# has 1 shared CPU at the most since a NUMA node has 4 CPUs and a pod is
+# requesting 1 CPU. Only one of the balloon that using NUMA node with
+#reserved CPU has 0 shared CPUs.
+CPUREQLIM="3"
+INITCPUREQLIM="100m-100m 100m-100m 100m-100m"
+POD_ANNOTATION="balloon.balloons.cri-resource-manager.intel.com: dynamic"
+n=8 create multicontainerpod
+sleep 5
```
Comment (jukkar) [inline test/e2e/policies.test-suite/balloons/n4c32/test01-dynamic-baloons/code.var.sh:13]: I have reservation about the forced sleeps, they just delay the progress of the test without much benefit. I think `verify-metrics-has-line` will try the fetch the stuff multiple times anyway. If not possible to remove the sleep, it should be made smaller.

File: test/e2e/policies.test-suite/balloons/n4c32/test01-dynamic-baloons/code.var.sh [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        len(nodes["pod3c0"]) == len(nodes["pod4c0"]) == len(nodes["pod5c0"]) == \
+        len(nodes["pod6c0"]) == len(nodes["pod7c0"]) == 1'
+
+# pod8: Add one more pod with 2 CPUs to inflate over NUMAs nodes, which should cross
+# the NUMA node boundaries but not the die boundaries. Because two NUMA nodes can offer
+# 2 CPUs in total. 
+CPUREQLIM="2"
+POD_ANNOTATION="balloon.balloons.cri-resource-manager.intel.com: dynamic"
+create multicontainerpod
+sleep 5
```
Comment (jukkar) [inline test/e2e/policies.test-suite/balloons/n4c32/test01-dynamic-baloons/code.var.sh:40]: same here

File: test/e2e/policies.test-suite/balloons/n4c32/test01-dynamic-baloons/code.var.sh [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+verify-metrics-has-line 'cpus_count="5"'
+verify-metrics-has-line 'sharedidlecpus="",sharedidlecpus_count="0"'
+verify-metrics-has-line 'sharedidlecpus_count="1"'
+verify 'len(nodes["pod8c0"])==2' \
+       'len(dies["pod8c0"])==1' \
+       'len(packages["pod8c0"])==1'
+kubectl delete pod pod8 --now
+sleep 8
```
Comment (jukkar) [inline test/e2e/policies.test-suite/balloons/n4c32/test01-dynamic-baloons/code.var.sh:48]: and same comment here too, and also the other sleeps below

File: test/e2e/policies.test-suite/balloons/n4c32/test01-dynamic-baloons/code.var.sh (hunk 1 above)
Comment (fmuyassarov) [reply test/e2e/policies.test-suite/balloons/n4c32/test01-dynamic-baloons/code.var.sh:13]: During local runs of these tests, I noticed that some metrics (like deletion of a pod and CPUs of it is becoming free) didn't get updated fast enough (default -metrics-interval 4 seconds) that caused some verification to fail. And similar issue was discovered by @askervin during his trials. And I just realized after your comment, that we don't need sleep for those checks that doesn't use verify-metrics-has-no-line and only verify-metrics-has-line. But for checks with verify-metrics-has-no-line we can either sleep here as it is now (should be more than default interval 4 secs) or I can add a timeout to verify-metrics-has-no-line as we have for [verify-metrics-has-no-line](https://github.com/intel/cri-resource-manager/blob/8a8ebb63b3e199ded81bd0e355b0d2ac56036d86/test/e2e/policies.test-suite/balloons/verify.source.sh#L7). WDYT?

Comment (askervin) [reply test/e2e/policies.test-suite/balloons/n4c32/test01-dynamic-baloons/code.var.sh:13]: @fmuyassarov , +1 from me to adding the timeout to verify-metrics-has-no-line, and getting rid of sleeps.

Comment (jukkar) [reply test/e2e/policies.test-suite/balloons/n4c32/test01-dynamic-baloons/code.var.sh:13]: Adding timeout to verify-metrics func sounds good, +1 to that.

**Review by klihub — CHANGES_REQUESTED** [review]
And when you address those review comments, please rebase this on latest master/HEAD. That should fix the build errors.

File: test/e2e/policies.test-suite/balloons/n4c32/test01-dynamic-baloons/code.var.sh (hunk 1 above)
Comment (fmuyassarov) [reply test/e2e/policies.test-suite/balloons/n4c32/test01-dynamic-baloons/code.var.sh]: done

**Comment by fmuyassarov** [comment]
> And when you address those review comments, please rebase this on latest master/HEAD. That should fix the build errors.

Thanks, done

**Review by jukkar — APPROVED** [review]
LGTM

**Review by askervin — DISMISSED** [review]
Thanks @fmuyassarov , LGTM.

**Review by jukkar — APPROVED** [review]

**Review by klihub — APPROVED** [review]

**Comment by fmuyassarov** [comment]
I don't have rights to merge it. If looks good, can someone please merge this patch?

**Comment by klihub** [comment]
@fmuyassarov Merged... but I already regret it a bit. I mean... ideally shouldn't we have waited for #932 getting merged before merging this ?

**Comment by fmuyassarov** [comment]
> @fmuyassarov Merged... but I already regret it a bit. I mean... ideally shouldn't we have waited for #931 getting merged before merging this ?

Yeah you are right, my bad :(
However, GitHub action doesn't run E2E test (I think) so even if we waited we can't run those tests on top of master but only locally as we do now. But, now we merged tests for improvements that don't exists yet on the master. Perhaps it is okay for now. If there are some functional changes that may happen on 931, then I will make sure to update the E2E tests accordingly.

Thanks for merging btw 😅 and I'm guilty if something comes up due to this

**Pull request merged**